src/QsCompiler/CompilerOptimizations/ConstantPropagation.fs

```diff
@@ -69,89 +68,54 @@ type ConstantPropagator(compiledCallables: ImmutableDictionary<QsQualifiedName,
                 let lhs = so.onSymbolTuple stm.Lhs
                 let rhs = so.ExpressionTransformation stm.Rhs
                 if stm.Kind = ImmutableBinding then
-                    stateRef := defineVarTuple !stateRef (lhs, rhs.Expression)
+                    constants <- defineVarTuple (shouldPropagate callables) constants (lhs, rhs.Expression)
                 QsBinding<TypedExpression>.New stm.Kind (lhs, rhs) |> QsVariableDeclaration
+
+            override this.onConditionalStatement stm =
+                let cbList, cbListEnd =
+                    stm.ConditionalBlocks |> Seq.fold (fun s (cond, block) ->
+                        let newCond = this.ExpressionTransformation cond
+                        match newCond.Expression with
+                        | BoolLiteral true -> s @ [None, block]
+                        | BoolLiteral false -> s
+                        | _ -> s @ [Some cond, block]
+                    ) [] |> List.ofSeq |> takeWhilePlus1 (fun (c, _) -> c <> None)
+                let newDefault = cbListEnd |> Option.map (snd >> Value) |? stm.Default
+
+                let cbList = cbList |> List.map (fun (c, b) -> this.onPositionedBlock (c, b))
+                let newDefault = match newDefault with Value x -> this.onPositionedBlock (None, x) |> snd |> Value | Null -> Null
+
+                match cbList, newDefault with
+                | [], Value x ->
+                    x.Body |> QsScopeStatement.New |> QsScopeStatement
+                | [], Null ->
+                    QsScope.New ([], LocalDeclarations.New []) |> QsScopeStatement.New |> QsScopeStatement
+                | _ ->
+                    let cases = cbList |> Seq.map (fun (c, b) -> (Option.get c, b))
+                    QsConditionalStatement.New (cases, newDefault) |> QsConditionalStatement
+
+            override this.onRepeatStatement stm =
+                constants <- enterScope constants
+                skipScope <- true
+                let result = base.onRepeatStatement stm
+                constants <- exitScope constants
+                result
+
+            override this.onQubitScope (stm : QsQubitScope) =
+                let kind = stm.Kind
+                let lhs = this.onSymbolTuple stm.Binding.Lhs
+                let rhs = this.onQubitInitializer stm.Binding.Rhs
+
+                jointFlatten (lhs, rhs) |> Seq.iter (fun (l, r) ->
+                    match l, r.Resolution with
+                    | VariableName name, QubitRegisterAllocation {Expression = IntLiteral num} ->
+                        let arrayIden = Identifier (LocalVariable name, Null) |> wrapExpr (ArrayType (ResolvedType.New Qubit))
+                        let elemI = fun i -> ArrayItem (arrayIden, IntLiteral (int64 i) |> wrapExpr Int)
+                        let expr = List.init (int num) (elemI >> wrapExpr Qubit) |> ImmutableArray.CreateRange |> ValueArray
```

Comment:
```suggestion
                        let expr = Seq.init (int num) (elemI >> wrapExpr Qubit) |> ImmutableArray.CreateRange |> ValueArray
```

Reply:
I don't think it really matters whether we use List or Seq in this case, but fixed in latest commit.